integrations/access/accesslist/app_test.go

```diff
@@ -181,6 +182,68 @@ func TestAccessListReminders(t *testing.T) {
 	}
 }
 
+type mockClient struct {
+	mock.Mock
+	teleport.Client
+}
+
+func (m *mockClient) ListAccessLists(ctx context.Context, pageSize int, pageToken string) ([]*accesslist.AccessList, string, error) {
+	args := m.Called(ctx, pageSize, pageToken)
+	return args.Get(0).([]*accesslist.AccessList), args.String(1), args.Error(2)
+}
+
+func TestAccessListReminders_BadClient(t *testing.T) {
+	t.Parallel()
+
+	clock := clockwork.NewFakeClockAt(time.Date(2023, 1, 1, 0, 0, 0, 0, time.UTC))
+
+	server, err := auth.NewTestServer(auth.TestServerConfig{
+		Auth: auth.TestAuthServerConfig{
+			Dir:   t.TempDir(),
+			Clock: clock,
+		},
+	})
+	require.NoError(t, err)
+	as := server.Auth()
+
+	// Use this mock client so that we can force ListAccessLists to return an error.
+	client := &mockClient{
+		Client: as,
+	}
+	client.On("ListAccessLists", mock.Anything, mock.Anything, mock.Anything).Return(([]*accesslist.AccessList)(nil), "", trace.BadParameter("error"))
+
+	bot := &mockMessagingBot{
+		recipients: map[string]*common.Recipient{
+			"owner1": {Name: "owner1"},
+			"owner2": {Name: "owner2"},
+		},
+	}
+	app := common.NewApp(&mockPluginConfig{client: client, bot: bot}, "test-plugin")
+	app.Clock = clock
+	ctx := context.Background()
+	go func() {
+		require.NoError(t, app.Run(ctx))
+	}()
+
+	ready, err := app.WaitReady(ctx)
+	require.NoError(t, err)
+	require.True(t, ready)
+
+	t.Cleanup(func() {
+		app.Terminate()
+		<-app.Done()
+		require.NoError(t, app.Err())
+	})
+
+	for i := 1; i <= 6; i++ {
+		clock.Advance(6 * time.Hour)
+		require.Eventuallyf(t, func() bool {
+			return len(client.Calls) == i
+		}, 5*time.Second, 5*time.Millisecond, "timeout waiting for expected number of calls (%d/%d)", len(client.Calls), i)
```

Comment:
5 millisecond retry period seems pretty quick for our slow CI environments. I would lean closer to 50 or 100.

But then again, if you use `clock.BlockUntil` correctly you might not need `Eventuallyf` at all, which is going to lead to a less flaky and more consistent test.

Reply:
So I was unaware of `BlockUntil` until now, but it looks like there's a latency that requires `Eventuallyf` even with `BlockUntil`, unfortunately. So I'll leave this as is, including the above comment.

Reply:
I'm not sure I agree that `Eventuallyf` is necessary.

Your reminder loop runs every `<-remindInterval.Next()`.

If you block until your loop is waiting for the tick, then set up the test conditions, then advance the clock the app loop will start to process.

Then you block again until the next `<-remindInterval.Next()`. At this point, you can be sure that the previous iteration of the reminder loop completed and can make assertions as to what should have happened.

Am I missing something?

Reply:
So, if I replace that wait logic with this:

```
	clock.BlockUntil(1)

	for i := 1; i <= 6; i++ {
		clock.Advance(6 * time.Hour)
		clock.BlockUntil(1)
		require.Len(t, client.Calls, i)
	}
```

When running the test locally, I get:
```
--- FAIL: TestAccessListReminders_BadClient (0.06s)
    /Users/mike/go/src/github.com/gravitational/teleport/integrations/access/accesslist/app_test.go:249: 
        	Error Trace:	/Users/mike/go/src/github.com/gravitational/teleport/integrations/access/accesslist/app_test.go:249
        	Error:      	"[]" should have 1 item(s), but has 0
        	Test:       	TestAccessListReminders_BadClient
```

However, if I add a `time.Sleep(1*time.Second)` before the `require.Len` check above, the test passes. So the only thing I can think of is that the call to count the number of calls is not synchronous for some reason? I'll dig a little more, but that's what I'm finding at the moment.